src/cache/gcs.rs

```diff
@@ -0,0 +1,353 @@
+            let mut request = Request::new(Method::Post, url.parse().unwrap());
+            {
+                let mut headers = request.headers_mut();
+                headers.set(Authorization(Bearer { token: creds.token }));
+                headers.set(ContentType("application/octet-stream".parse().unwrap()));
+                headers.set(ContentLength(content.len() as u64));
+            }
+            request.set_body(content);
+
+            client.request(request).then(|result| {
+                match result {
+                    Ok(res) => {
+                        if res.status().is_success() {
+                            trace!("PUT succeeded");
+                            Ok(())
+                        } else {
+                            trace!("PUT failed with HTTP status: {}", res.status());
+                            Err(ErrorKind::BadHTTPStatus(res.status().clone()).into())
+                        }
+                    }
+                    Err(e) => {
+                        trace!("PUT failed with error: {:?}", e);
+                        Err(e.into())
+                    }
+                }
+            })
+        }))
+    }
+}
+
+pub struct GCSCredentialProvider {
+    read_only: bool,
+    credentials_path: String,
+    cached_credentials: RefCell<Option<Shared<SFuture<GCSCredential>>>>,
+}
+
+#[derive(Debug, Deserialize)]
+struct ServiceAccountKey {
+    #[serde(rename = "type")]
+    _type: String,
+    project_id: String,
+    private_key_id: String,
+    private_key: String,
+    client_email: String,
+    client_id: String,
+    auth_uri: String,
+    token_uri: String,
+    auth_provider_x509_cert_url: String,
+}
+
+#[derive(Serialize)]
+struct JwtClaims {
+    #[serde(rename = "iss")]
+    issuer: String,
+    scope: String,
+    #[serde(rename = "aud")]
+    audience: String,
+    #[serde(rename = "exp")]
+    expiration: i64,
+    #[serde(rename = "iat")]
+    issued_at: i64,
+}
+
+#[derive(Deserialize)]
+struct TokenMsg {
+    access_token: String,
+    token_type: String,
+    expires_in: usize,
+}
+
+#[derive(Clone)]
+pub struct GCSCredential {
+    token: String,
+    expiration_time: chrono::DateTime<chrono::UTC>,
+}
+
+impl GCSCredentialProvider {
+    pub fn new(read_only: bool, credentials_path: String) -> Self {
+        GCSCredentialProvider {
+            read_only,
+            credentials_path,
+            cached_credentials: RefCell::new(None),
+        }
+    }
+
+    fn auth_request_jwt(&self, expire_at: &chrono::DateTime<chrono::UTC>) -> Result<String> {
+        let metadata = fs::metadata(&self.credentials_path).chain_err(|| {
+            "Couldn't stat GCS credentials file"
+        })?;
+        if !metadata.is_file() {
+            bail!("Couldn't open GCS credentials file.");
+        }
+        let mut file = File::open(&self.credentials_path)?;
+        let mut service_account_json = String::new();
+        file.read_to_string(&mut service_account_json)?;
```

Comment:
SGTM-- I only did it this way because I was trying to match the behavior for S3, which reads in the files during the event loop, just before it goes to use them. See [here](https://github.com/mozilla/sccache/blob/master/src/simples3/credential.rs#L187).